docs/puzzles/day4.md

```diff
@@ -1,11 +1,232 @@
+```scala
+val numbers: List[Int] = inputSections.head.split(',').map(_.toInt)
+```
+
+### Parsing Boards
+
+A board is just a table of integers, and a table is just a list of lines, where each line is also list.
+And so we have our type for the boards: `List[List[Int]]`!
+
+But not so fast, we would like to add some extra operations on boards, so we wrap it in a [case classe](https://docs.scala-lang.org/tour/case-classes.html):
+(don't worry if you don't understand the methods, we'll explain them later)
+```scala
+case class Board(lines: List[List[Int]]):
+  def mapNumbers(f: Int => Int): Board = Board(lines.map(_.map(f)))
+  def columns: List[List[Int]] = lines.transpose
+```
+
+Now that we have our representation, we still have to actually parse the string that represents a board, for that we'll create a [companion object](https://docs.scala-lang.org/scala3/book/taste-objects.html) with a method `parse` that takes a `String` as input, and returns a `Board`:
+
+```scala
+object Board:
+  def parse(inputBoard: String): Board = ???
+```
+
+Let's start easy, assuming we have a line, how do we find all the numbers?
+We can use a [regular expressions](https://en.wikipedia.org/wiki/Regular_expression)(Regexes): 
+ - "digit" -> `\d`
+ - "one or more" -> `+`
+ - "one or more digits" -> `\d+`
+
+```scala
+val numberParser = raw"\d+".r
+```
+`raw` allows us to write things like `\d` without it being translated as a line return.
+`.r` converts our `String` to a `Regex`.
+
+For each line, `numberParser` finds every number in it, and we parse them to `Int`:
+```scala
+def parseLine(inputLine: String): List[Int] =
+  numberParser.findAllIn(inputLine).toList.map(_.toInt)
+```
+And the lines are just separated by newlines:
+```scala
+val lines = inputBoard.split('\n').toList
+Board(lines.map(parseLine))
+```
+
+Where `lines.map(parseLine)` means:
+1. create a new list
+2. for each `line` in `lines` put `parseLine(line)` in the list
+
+This gives us a `List[List[Int]]`, which we use to construct a `Board`.
+
+Since we have multiple `Board`s:
+```scala
+val originalBoards: List[Board] = inputSections.tail.map(Board.parse)
+```
+
+## Reasonning about the problem
+
+It's kind of difficult to think about all those numbers being picked at random turn.
+We can simplify the problem by replacing each number by the "turn" at which it was drawn.
+
+`zipWithIndex` transforms our list of numbers into a list of number-index pair, where the index is just the turn at which the number is picked.
+We can then convert it to a `Map`, to be able to use it like a function.
+To be able also to go back, we invert our `Map` by swapping its keys and values.
+
+```scala
+val numberToTurn = numbers.zipWithIndex.toMap
+val turnToNumber = numberToTurn.map((number -> turn) => (turn -> number))
+```
+
+Our simplified boards are therefore:
+```scala
+val boards = originalBoards.map(board => board.mapNumbers(numberToTurn))
```

Comment:
You said earlier that you would explain what `mapNumbers` is, but you haven't done so.